File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:

        def find_count(root: TreeNode, value: int) -> int:
            if not root:
                return 0
            
            res = 0
            if root.val >= value:
                res += 1
            
            res += find_count(root.left, max(value, root.val))
            res += find_count(root.right, max(value, root.val))
            return res

        return find_count(root, root.val);

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<int, int> ump;
    int index = 0;
    TreeNode* actual_treebuild(vector<int>& preorder, int l, int r) {
        if(l > r) return nullptr;
        int val = preorder[index++];
        TreeNode* root = new TreeNode(val);
        root->left = actual_treebuild(preorder, l, ump[val]-1);
        root->right = actual_treebuild(preorder, ump[val]+1, r);
        return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int n = inorder.size();
        for(int i=0;i<n;i++) {
            ump[inorder[i]] = i;
        }
        return actual_treebuild(preorder, 0, n-1);
    }
};